File: frontend/src/api/accounts.ts
import { api } from "./client";
import type { Account, AccountInput } from "../types/account";

export async function listAccounts() {
  const response = await api.get<unknown>("/accounts");
  return assertAccounts(response.data);
}

export async function getAccount(accountId: string) {
  const response = await api.get<unknown>(`/accounts/${accountId}`);
  return assertAccount(response.data);
}

export async function createAccount(input: AccountInput) {
  const response = await api.post<Account>("/accounts", input);
  return response.data;
}

export async function updateAccount(accountId: string, input: Partial<AccountInput>) {
  const response = await api.patch<Account>(`/accounts/${accountId}`, input);
  return response.data;
}

export async function deleteAccount(accountId: string) {
  await api.delete(`/accounts/${accountId}`);
}

export async function syncAccountProfile(accountId: string) {
  const response = await api.post<Account>(`/accounts/${accountId}/sync-profile`);
  return response.data;
}

export async function createAccountSession(accountId: string) {
  const response = await api.post<Account>(`/accounts/${accountId}/session/create`);
  return response.data;
}

export async function finishAccountSession(accountId: string) {
  const response = await api.post<Account>(`/accounts/${accountId}/session/finish`);
  return response.data;
}

export async function validateAccountSession(accountId: string) {
  const response = await api.post<Account>(`/accounts/${accountId}/session/validate`);
  return response.data;
}

export async function refreshAccountSession(accountId: string) {
  const response = await api.post<Account>(`/accounts/${accountId}/session/refresh`);
  return response.data;
}

export async function deleteAccountSession(accountId: string) {
  const response = await api.delete<Account>(`/accounts/${accountId}/session`);
  return response.data;
}

export async function openAccountBrowser(accountId: string) {
  const response = await api.post<Account>(`/accounts/${accountId}/browser/open`);
  return response.data;
}

export async function openAccountHome(accountId: string) {
  const response = await api.post<Account>(`/accounts/${accountId}/browser/open-home`);
  return response.data;
}

function assertAccounts(value: unknown): Account[] {
  if (!Array.isArray(value)) {
    throw new Error("Account list response was not JSON array data.");
  }
  return value.map(assertAccount);
}

function assertAccount(value: unknown): Account {
  if (!isRecord(value)) {
    throw new Error("Account response was not JSON object data.");
  }
  if (
    typeof value.id !== "string" ||
    typeof value.nickname !== "string" ||
    typeof value.username !== "string" ||
    value.platform !== "reddit" ||
    !["active", "paused", "error", "archived"].includes(String(value.status))
  ) {
    throw new Error("Account response did not match the frontend account contract.");
  }
  return value as unknown as Account;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

File: backend/app/api/v1/accounts.py
from uuid import UUID

from fastapi import APIRouter, Depends, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_session
from app.schemas.account import AccountCreate, AccountLogin, AccountRead, AccountUpdate
from app.services.account_service import AccountService
from app.services.browser_session_service import BrowserSessionService

router = APIRouter(prefix="/accounts", tags=["accounts"])


def service(session: AsyncSession = Depends(get_session)) -> AccountService:
    return AccountService(session)


def browser_session_service(session: AsyncSession = Depends(get_session)) -> BrowserSessionService:
    return BrowserSessionService(session)


@router.get("", response_model=list[AccountRead])
async def list_accounts(
    search: str | None = None,
    account_service: AccountService = Depends(service),
) -> list[AccountRead]:
    return await account_service.list_accounts(search)


@router.post("", response_model=AccountRead, status_code=status.HTTP_201_CREATED)
async def create_account(
    payload: AccountCreate,
    account_service: AccountService = Depends(service),
) -> AccountRead:
    return await account_service.create_account(payload)


@router.get("/{account_id}", response_model=AccountRead)
async def get_account(
    account_id: UUID,
    account_service: AccountService = Depends(service),
) -> AccountRead:
    return await account_service.get_account(account_id)


@router.patch("/{account_id}", response_model=AccountRead)
async def update_account(
    account_id: UUID,
    payload: AccountUpdate,
    account_service: AccountService = Depends(service),
) -> AccountRead:
    return await account_service.update_account(account_id, payload)


@router.delete("/{account_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_account(
    account_id: UUID,
    account_service: AccountService = Depends(service),
) -> Response:
    await account_service.delete_account(account_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/{account_id}/login", response_model=AccountRead)
async def login_account(
    account_id: UUID,
    payload: AccountLogin | None = None,
    session_service: BrowserSessionService = Depends(browser_session_service),
) -> AccountRead:
    return await session_service.login(account_id, payload)


@router.post("/{account_id}/session/validate", response_model=AccountRead)
async def validate_account_session(
    account_id: UUID,
    session_service: BrowserSessionService = Depends(browser_session_service),
) -> AccountRead:
    return await session_service.validate(account_id)


@router.post("/{account_id}/session/refresh", response_model=AccountRead)
async def refresh_account_session(
    account_id: UUID,
    session_service: BrowserSessionService = Depends(browser_session_service),
) -> AccountRead:
    return await session_service.refresh(account_id)


@router.delete("/{account_id}/session", response_model=AccountRead)
async def delete_account_session(
    account_id: UUID,
    session_service: BrowserSessionService = Depends(browser_session_service),
) -> AccountRead:
    return await session_service.delete(account_id)


@router.post("/{account_id}/session/logout", response_model=AccountRead)
async def logout_account_session(
    account_id: UUID,
    session_service: BrowserSessionService = Depends(browser_session_service),
) -> AccountRead:
    return await session_service.logout(account_id)


@router.post("/{account_id}/browser/open", response_model=AccountRead)
async def open_account_browser(
    account_id: UUID,
    session_service: BrowserSessionService = Depends(browser_session_service),
) -> AccountRead:
    return await session_service.open_browser(account_id)


@router.post("/{account_id}/browser/open-home", response_model=AccountRead)
async def open_account_home(
    account_id: UUID,
    session_service: BrowserSessionService = Depends(browser_session_service),
) -> AccountRead:
    return await session_service.open_home(account_id)
